essay/swar/latency.md

```diff
@@ -0,0 +1,105 @@
+```c++
+uint32_t parse_eight_digits_swar(const char *chars) {
+  uint64_t val;
+  memcpy(&val, chars, 8);
+  val = val - 0x3030303030303030;
+  uint64_t byte10plus   = ((val        
+      * (1 + (0xa  <<  8))) >>  8)
+      & 0x00FF00FF00FF00FF;
+  uint64_t short100plus = ((byte10plus 
+       * (1 + (0x64 << 16))) >> 16) 
+       & 0x0000FFFF0000FFFF;
+  short100plus *= (1 + (10000ULL << 32));
+  return short100plus >> 32;
+}
+```
+
+Lemire commented:
+
+> The fact that each step depends on the completion of the previous step is less good: it means that the latency is relatively high
+
+Certainly this implementation is a long chain of data dependencies, or deep data dependencies, but I still think Lemire's comment is unfortunate, the main topic of this essay.
+
+By the way, we [implemented this code using SWAR](https://github.com/thecppzoo/zoo/blob/9a1e406c63ede7cbc16cf16bde3731fcf9ee8c86/benchmark/atoi.cpp#L33-L55):
+```c++
+uint32_t calculateBase10(zoo::swar::SWAR<8, uint64_t> convertedToIntegers) noexcept {
+    /* the idea is to perform the following multiplication:
+     * NOTE: THE BASE OF THE NUMBERS is 256 (2^8), then 65536 (2^16), 2^32
+     * convertedToIntegers is IN BASE 256 the number ABCDEFGH
+     * BASE256:   A    B    C    D    E    F    G    H *
+     * BASE256                                 10    1 =
+     *           --------------------------------------
+     * BASE256  1*A  1*B  1*C  1*D  1*E  1*F  1*G  1*H +
+     * BASE256 10*B 10*C 10*D 10*E 10*F 10*G 10*H    0
+     *           --------------------------------------
+     * BASE256 A+10B ....................... G+10H   H
+     * See that the odd-digits (base256) contain 10*odd + even
+     * Then, we can use base(2^16) digits, and base(2^32) to
+     * calculate the conversion for the digits in 100s and 10,000s
+     */
+    auto by11base256 = convertedToIntegers.multiply(256*10 + 1);
+    auto bytePairs = zoo::swar::doublePrecision(by11base256).odd;
+    static_assert(std::is_same_v<decltype(bytePairs), zoo::swar::SWAR<16, uint64_t>>);
+    auto by101base2to16 = bytePairs.multiply(1 + (100 << 16));
+    auto byteQuads = zoo::swar::doublePrecision(by101base2to16).odd;
+    auto by10001base2to32 = byteQuads.multiply(1 + (10000ull << 32));
+    return uint32_t(by10001base2to32.value() >> 32);
+}
+```
+resulting in identical code generation even with minimal optimization, and we will keep iterating so that the code reflects this is just one example of what we call "associative iteration" --[Glossary](../../glossary.md#associative-iteration)--.
+
+## Parallelism
+
+The essence of parallel algorithms is the "mapping" of an operation that is associative to multiple data elements simultaneously or in parallel, and the collection (reduction, gathering, combination) of the partial results into the final result.
+
+The combination of the partial results induce a tree structure, whose height is in the order of the logarithm of the number of elements N.
```

Comment:
maybe outside the scope of what you want to do, but possibly a diagram of the "tree structure" could be illuminating?